### PR TITLE
[Next ride event] add event settings

### PR DESCRIPTION
**Pull request by mltbnz**
## 📲 What

The PR started to add a settings screen for the event feature where users can configure what kind of event notifications they want to see in the app and also if at all they want to see event notifications. Additionally the PR reworks the before called `Section` object to be reusable for others screens and also the `SettingsViewController` can now be subclassed. To configure a SettingsViewController you now inject and Array of `SettingsSection`. The before `SettingsViewController` is now replicated and called `AppSettingsViewController` and the Event settings can be found in the `EventSettingsViewController`.

+ The settings are being saved in the UserDefaults.
+ NextRideManager is injected with the settings to configure the feature.
+ `RideType` was addd to the `Ride` model  

## 🤔 Why

To let users configure the next ride feature. 

## 👀 See

Screenshots, external resources?
- [x] Added snapshot tests (using iPhone 8 Simulator)

## ♿️ Accessibility 

- [x] Tap targets use minimum of 44x44 pts dimensions
- [x] Works with VoiceOver
- [x] Supports Dynamic Type 

## ⏰ TODO

+ Further refactor `SettingsSection` to be more declaritive

### CODE REVIEW

File: CriticalMass/AppSettingsViewController.swift [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    override func createViewController(type: UIViewController.Type) -> UIViewController {
+        // Perform dependency injection if needed
+        switch type {
+        case _ as ManageFriendsViewController.Type:
+            return ManageFriendsViewController(dataStore: dataStore, idProvider: idProvider)
+        case _ as ThemeSelectionViewController.Type:
+            return ThemeSelectionViewController(themeController: themeController)
+        case _ as EventSettingsViewController.Type:
+            return EventSettingsViewController(
+                controllerTitle: "Event Settings",
```
Comment (fbernutz) [inline CriticalMass/AppSettingsViewController.swift:54]: Shouldn't "Event Settings" be localized? 🤔

Comment (fbernutz) [reply CriticalMass/AppSettingsViewController.swift:54]:
```suggestion
                controllerTitle: L10n.settingsEventSettings",
```

File: CriticalMass/NextRideManager.swift [hunk 2]
```diff
@@ -5,59 +5,85 @@ import CoreLocation
...
         case let .success(rides):
-            let rangeFilteredRides = filterRidesInRange(rides, userCoordinate)
-            guard !rangeFilteredRides.isEmpty else {
+            guard !rides.isEmpty else {
                 handler(.failure(EventError.rideIsOutOfRangeError))
                 return
             }
-            guard let ride = getUpcomingRide(rangeFilteredRides) else {
+            if rides.compactMap(\.rideType).isEmpty {
+                // All rides do not have a rideType and so the filtering is skipped
```
Comment (fbernutz) [inline CriticalMass/NextRideManager.swift:70]: Maybe the comment would be better to understand the other way around:
```suggestion
                // None of the rides have a rideType and so the filtering is skipped
```

File: CriticalMapsSnapshotTests/AppSettingsViewControllerSnapshotTests.swift [hunk 3]
```diff
@@ -3,8 +3,7 @@
 
 @testable import CriticalMaps
 import XCTest
-
-class SettingsViewControllerSnapshotTests: XCTestCase {
+class AppSettingsViewControllerSnapshotTests: XCTestCase {
```
Comment (fbernutz) [inline CriticalMapsSnapshotTests/AppSettingsViewControllerSnapshotTests.swift:6]: When renaming the class of a SnapshotTest, there'll be new snapshot references. So we have to clean up the old references manually. There are right now old `SettingsViewControllerSnapshotTests` references in the project. Could you please clean them up?

File: CriticalMass/SettingsCheckTableViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,20 @@
+//
+// Created for CriticalMaps in 2020
+
+import UIKit
+
+class SettingsCheckTableViewCell: UITableViewCell, IBConstructable {
+    private var switchable: Toggleable?
```
Comment (fbernutz) [inline CriticalMass/SettingsCheckTableViewCell.swift:7]: switchable or toggleable? hm, kind of confusing...

File: CriticalMass/SettingsSection.swift [hunk 5]
```diff
@@ -8,126 +8,204 @@
 
 import UIKit
 
-enum Section: CaseIterable {
-    typealias AllCases = [Section]
+/** TODO: Refactor to a more versataile API like:
+Section(title: "NAME", rows: [SwitchRow(title:subTitle:actionClosure:), OptionRow]).....
```
Comment (fbernutz) [inline CriticalMass/SettingsSection.swift:12]: I think this will be done in a separate task, right?

**Comment by fbernutz** [comment]
Another thing I recognized in the EventSettingsVC (in the snapshot reference as well and in the AppSettingsVC on "Info" Section), there seems to be a color issue on the tableView Section Titles in iOS 14 Simulator. I can't see this on my device (iOS 14) with the CM App from the Store. 

simulator_screenshot_40DA27B1-EAC4-4B11-9CF0-2B3EBA03E5F5

Should I open an issue for that?

File: CriticalMass/SettingsCheckTableViewCell.swift (hunk 4 above)
Comment (mltbnz) [reply CriticalMass/SettingsCheckTableViewCell.swift:7]: haha. true. actually this is not needed and I have a fix PR already in place #343

File: CriticalMass/SettingsSection.swift (hunk 5 above)
Comment (mltbnz) [reply CriticalMass/SettingsSection.swift:12]: Yes. Just wanted to remind me of it.

File: CriticalMass/SettingsCheckTableViewCell.swift (hunk 4 above)
Comment (mltbnz) [reply CriticalMass/SettingsCheckTableViewCell.swift:7]: So I will add a TODO and then clean that up after #343 has landed

**Comment by mltbnz** [comment]
> Another thing I noticed in the EventSettingsVC (in the snapshot reference as well and in the AppSettingsVC on "Info" Section), there seems to be a color issue on the tableView Section Titles in iOS 14 Simulator. I can't see this on my device (iOS 14) with the CM App from the Store.
> 
> simulator_screenshot_40DA27B1-EAC4-4B11-9CF0-2B3EBA03E5F5
> 
> Should I open an issue for that?

That was a little pain and the `Theme` implementation came back to bit me in the #!+@ again. I was not able to figure out where we set the color for it and so I just set 1 color for both themes. Do you maybe know?

File: CriticalMass/SettingsCheckTableViewCell.swift (hunk 4 above)
Comment (mltbnz) [reply CriticalMass/SettingsCheckTableViewCell.swift]: Updated

**Comment by fbernutz** [comment]
> > Another thing I noticed in the EventSettingsVC (in the snapshot reference as well and in the AppSettingsVC on "Info" Section), there seems to be a color issue on the tableView Section Titles in iOS 14 Simulator. I can't see this on my device (iOS 14) with the CM App from the Store.
> > Should I open an issue for that?
> 
> That was a little pain and the `Theme` implementation came back to bit me in the #!+@ again. I was not able to figure out where we set the color for it and so I just set 1 color for both themes. Do you maybe know?

I hope you don't mind that I pushed in this branch directly. I fixed the color issue and recorded the new Snapshots. It had something to do with a View inside the Section Header. 
I saw that the translations for the event strings are currently only present in the english version. Do you want to add them to the other languages as well before merging?

**Comment by mltbnz** [comment]
> > > Another thing I noticed in the EventSettingsVC (in the snapshot reference as well and in the AppSettingsVC on "Info" Section), there seems to be a color issue on the tableView Section Titles in iOS 14 Simulator. I can't see this on my device (iOS 14) with the CM App from the Store.
> > > Should I open an issue for that?
> > 
> > 
> > That was a little pain and the `Theme` implementation came back to bit me in the #!+@ again. I was not able to figure out where we set the color for it and so I just set 1 color for both themes. Do you maybe know?
> 
> I hope you don't mind that I pushed in this branch directly. I fixed the color issue and recorded the new Snapshots. It had something to do with a View inside the Section Header.
> I saw that the translations for the event strings are currently only present in the english version. Do you want to add them for the other languages as well before merging?

I could do that but maybe that is also a task for another PR? I would update the Crowdin resource to sync it with the other translations. Would also then be updating the localisations from #330

**Comment by fbernutz** [comment]
Fine for me :+1: 
Ah damn, I recorded the Snapshots for iOS 14. Updated them, hope the CI is green now. 
Do you have anything to add or can we then merge this? :)

**Review by fbernutz — APPROVED** [review]

**Comment by mltbnz** [comment]
Thanks.
I'm alright for now.

Will work on the Select Cell refactor and translations update in the next PRs

**Pull request merged**